### PR TITLE
feat: Force users to change their passwords after 3/6/12 months

### PR DESCRIPTION
**Pull request by ljl66-66**
Fix: https://github.com/casdoor/casdoor/issues/3330

### CODE REVIEW

**Comment by ljl66-66** [comment]
source of issue：[#issue3330](https://github.com/casdoor/casdoor/issues/3330)

File: object/check.go [hunk 1]
```diff
@@ -382,6 +382,11 @@ func CheckUserPassword(organization string, username string, password string, la
 			return nil, err
 		}
 	}
+	if CheckPasswordExpired(user) {
+		user.NeedUpdatePassword = true
+		UpdateUser(user.GetId(), user, []string{"need_update_password"}, false)
+		return nil, fmt.Errorf(i18n.Translate(lang, "check:Your password has not been changed for too long, please click Forgot Password to change your password"))
```
Comment (hsluoyz) [inline object/check.go:388]: use ChatGPT, don't write English text by yourself

File: object/check.go [hunk 2]
```diff
@@ -382,6 +382,11 @@ func CheckUserPassword(organization string, username string, password string, la
 			return nil, err
```
Comment (hsluoyz) [inline object/check.go:382]: Check in non-LDAP path

File: object/check.go [hunk 3]
```diff
@@ -382,6 +382,11 @@ func CheckUserPassword(organization string, username string, password string, la
...
 		}
 	}
+	if CheckPasswordExpired(user) {
+		user.NeedUpdatePassword = true
```
Comment (hsluoyz) [inline object/check.go:386]: Don't mess with NeedUpdatePassword if you don't fully understand and know how to use it

File: object/check_password_expired.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
+package object
+
+import (
+	"time"
+
+	"github.com/casvisor/casvisor-go-sdk/casvisorsdk"
+)
+
+func CheckPasswordExpired(user *User) bool {
```
Comment (hsluoyz) [inline object/check_password_expired.go:9]: C -> c

File: object/check_password_expired.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	organization, err := GetOrganizationByUser(user)
+	if err != nil {
+		return false
+	}
+	passwordExpireDays := organization.PasswordExpireDays
+	if passwordExpireDays <= 0 {
+		return false
+	}
+	lastChangeTime, err := getLastTimeOfAction(organization.Name, user.Name, "set-password")
```
Comment (hsluoyz) [inline object/check_password_expired.go:18]: Add a field to user to record the last password change time, don't iterate on logs

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: object/check.go [hunk 6]
```diff
@@ -382,6 +382,11 @@ func CheckUserPassword(organization string, username string, password string, la
...
+		UpdateUser(user.GetId(), user, []string{"need_update_password"}, false)
```
Comment (hsluoyz) [inline object/check.go:387]: We handle all errors returned by UpdateUser

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: object/user.go [hunk 7]
```diff
@@ -203,10 +203,11 @@ type User struct {
...
 
-	ManagedAccounts    []ManagedAccount `xorm:"managedAccounts blob" json:"managedAccounts"`
-	MfaAccounts        []MfaAccount     `xorm:"mfaAccounts blob" json:"mfaAccounts"`
-	NeedUpdatePassword bool             `json:"needUpdatePassword"`
-	IpWhitelist        string           `xorm:"varchar(200)" json:"ipWhitelist"`
+	ManagedAccounts        []ManagedAccount `xorm:"managedAccounts blob" json:"managedAccounts"`
+	MfaAccounts            []MfaAccount     `xorm:"mfaAccounts blob" json:"mfaAccounts"`
+	NeedUpdatePassword     bool             `json:"needUpdatePassword"`
+	IpWhitelist            string           `xorm:"varchar(200)" json:"ipWhitelist"`
+	LastChangePasswordTime string           `xorm:"varchar(100)" json:"lastChangePasswordTime"`
```
Comment (hsluoyz) [inline object/user.go:210]: Put before LastSigninWrongTime

File: object/user.go [hunk 8]
```diff
@@ -697,7 +698,7 @@ func UpdateUser(id string, user *User, columns []string, isAdmin bool) (bool, er
 			"eveonline", "fitbit", "gitea", "heroku", "influxcloud", "instagram", "intercom", "kakao", "lastfm", "mailru", "meetup",
 			"microsoftonline", "naver", "nextcloud", "onedrive", "oura", "patreon", "paypal", "salesforce", "shopify", "soundcloud",
 			"spotify", "strava", "stripe", "type", "tiktok", "tumblr", "twitch", "twitter", "typetalk", "uber", "vk", "wepay", "xero", "yahoo",
-			"yammer", "yandex", "zoom", "custom", "need_update_password", "ip_whitelist",
+			"yammer", "yandex", "zoom", "custom", "need_update_password", "ip_whitelist", "last_change_password_time",
```
Comment (hsluoyz) [inline object/user.go:701]: Put before LastSigninWrongTime

File: web/src/UserEditPage.js [hunk 9]
```diff
@@ -1083,6 +1083,19 @@ class UserEditPage extends React.Component {
           </Col>
         </Row>
       );
+    } else if (accountItem.name === "Last change password time") {
+      return (
+        <Row style={{marginTop: "20px"}} >
+          <Col style={{marginTop: "5px"}} span={(Setting.isMobile()) ? 22 : 2}>
+            {Setting.getLabel(i18next.t("user:Last change password time"), i18next.t("user:Last change password time"))} :
```
Comment (hsluoyz) [inline web/src/UserEditPage.js:1090]: Put before LastSigninWrongTime

File: controllers/user.go [hunk 10]
```diff
@@ -561,8 +562,9 @@ func (c *ApiController) SetPassword() {
 	targetUser.Password = newPassword
 	targetUser.UpdateUserPassword(organization)
 	targetUser.NeedUpdatePassword = false
+	targetUser.LastChangePasswordTime = time.Now().Format(time.RFC3339)
```
Comment (hsluoyz) [inline controllers/user.go:565]: use util.GetCurrentTime

File: object/check_password_expired.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
+package object
+
+import (
+	"time"
+)
+
+func checkPasswordExpired(user *User) bool {
```
Comment (hsluoyz) [inline object/check_password_expired.go:7]: return error like CheckPassword

File: object/check_password_expired.go [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+	organization, err := GetOrganizationByUser(user)
+	if err != nil {
+		return false
+	}
```
Comment (hsluoyz) [inline object/check_password_expired.go:11]: check nil for organization

File: object/check_password_expired.go [hunk 13]
```diff
@@ -0,0 +1,26 @@
+package object
```
Comment (hsluoyz) [inline object/check_password_expired.go:1]: Apache header

File: object/check_password_expired.go [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+	if err != nil {
+		return false
+	}
+	passwordExpireDays := organization.PasswordExpireDays
+	if passwordExpireDays <= 0 {
+		return false
+	}
+	lastChangePasswordTime := user.LastChangePasswordTime
+	if lastChangePasswordTime == "" {
+		return false
```
Comment (hsluoyz) [inline object/check_password_expired.go:18]: If I didn't ever change password, I can bypass this check, is this OK? It's a security hole!

You should use user created time if lastChangePasswordTime is empty

If user created time is also empty, then return error

File: object/check_password_expired.go [hunk 15]
```diff
@@ -1,26 +1,53 @@
...
-	if err != nil {
-		return false
+		if user.CreatedTime == "" {
+			return fmt.Errorf(i18n.Translate(lang, "check:Your password has expired. Please reset your password by clicking \"Forgot password\""))
+		}
+		lastChangePasswordTime = user.CreatedTime
 	}
+
+	lastTime := util.String2Time(lastChangePasswordTime)
 	expiryDate := lastTime.AddDate(0, 0, passwordExpireDays)
```
Comment (hsluoyz) [inline object/check_password_expired.go:48]: expiryDate -> expireTime

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Pull request merged**